### PR TITLE
Fix: Prevent Path Traversal / Arbitrary File Read in nltk.data when resource omits protocol

### LINKED ISSUE #3488: Path Traversal via No‑Protocol Resources
#  Path Traversal via No-Protocol Resources

## Summary
Resource names without explicit protocols (e.g., `"../../etc/passwd"`) could resolve as filesystem paths, enabling arbitrary local file read attacks.

## Root Cause
- No-protocol names were implicitly treated as `nltk:` resources
- Traversal segments (`../`), absolute paths, and Windows drive letters were not properly rejected
- Some API callers could bypass normalization and access filesystem paths directly

## Affected Components
- `nltk.data.normalize_resource_url()`
- `nltk.data.find()`
- Zipfile path parsing in `nltk.data`
- Related resource loading code paths

## Security Impact
- Arbitrary local file read via path traversal
- Potential exposure of sensitive system files
- Attack persists even with API misuse

## Fix Implemented
- Added regex validation to reject unsafe patterns (`../`, absolute paths, `C:\`, backslashes)
- Implemented defense-in-depth validation in `find()` method
- Maintained compatibility with valid `nltk:` and `file:` URLs
- Added comprehensive unit tests

## Resolution
**Fixed in PR #3467** – *Prevent Path Traversal / Arbitrary File Read in nltk.data*

### PR DESCRIPTION
**Pull request by HyperPS**
Fix: #3488 
Summary
-------
This patch closes a path traversal / arbitrary file-read vector in nltk.data
where a no-protocol resource string containing traversal segments (e.g. "../../etc/passwd")
could be resolved against an empty path entry and returned as a FileSystemPathPointer.

What changed
-----------
- Add _UNSAFE_NO_PROTOCOL_RE and _reject_unsafe_no_protocol() to validate inputs that omit a protocol.
- Validate no-protocol inputs in normalize_resource_url() before treating them as 'nltk:' resources.
- Add a defense-in-depth check in find() to reject traversal-like names when find() is called directly.
- Fix zipfile regex matching to avoid accidental captures.
- Add pytest tests to assert traversal / absolute names are rejected.

Security impact
---------------
Prevents reading arbitrary local files via no-protocol resource strings. Callers should use 'nltk:' for package resources and 'file:' for explicit filesystem paths.

### CODE REVIEW

**Comment by ekaf** [comment]
Thanks @HyperPS, but first you'll need to follow the guidelines concerning pre-commit. Please see https://github.com/nltk/nltk/pull/3461#issuecomment-3518251102

**Comment by HyperPS** [comment]
pre-commit done.

**Comment by ekaf** [comment]
The security fix is useful,  but the implementation feels quite heavy for what it is trying to achieve. It might be possible to express the same checks in a more concise and readable way.

**Comment by HyperPS** [comment]
Thanks @ekaf — I’ve pushed updates to simplify the validation logic as suggested.
The checks for traversal, absolute paths, and Windows drive letters are now handled in a smaller, more readable helper while keeping the same security guarantees.
Also, since this affects path resolution and can lead to unintended file access, I reported it through GitHub’s Security Advisory panel as well.
When possible, please review and accept the advisory so it can be tracked properly.

**Comment by HyperPS** [comment]
Thanks, @ekaf — the path traversal fix is now fully stable.
• No-protocol resource names are validated with a simple, readable helper
• Traversal, absolute paths, and Windows drive letters are correctly rejected
• Behavior is unchanged for valid 'nltk:' and 'file:' URLs
• downloader.py restored so this PR stays focused on nltk.data
• mapping.py update included only to fix Windows CI as discussed

Let me know if any adjustments are needed.

File: nltk/data.py [hunk 1]
```diff
@@ -144,58 +145,35 @@ def split_resource_url(resource_url):
 
 
 def normalize_resource_url(resource_url):
```
Comment (ekaf) [inline nltk/data.py:175]: The original docstring is still relevant:

```suggestion
def normalize_resource_url(resource_url):
    r"""
    Normalizes a resource url

    >>> windows = sys.platform.startswith('win')
    >>> os.path.normpath(split_resource_url(normalize_resource_url('file:grammar.fcfg'))[1]) == \
    ... ('\\' if windows else '') + os.path.abspath(os.path.join(os.curdir, 'grammar.fcfg'))
    True
    >>> not windows or normalize_resource_url('file:C:/dir/file') == 'file:///C:/dir/file'
    True
    >>> not windows or normalize_resource_url('file:C:\\dir\\file') == 'file:///C:/dir/file'
    True
    >>> not windows or normalize_resource_url('file:C:\\dir/file') == 'file:///C:/dir/file'
    True
    >>> not windows or normalize_resource_url('file://C:/dir/file') == 'file:///C:/dir/file'
    True
    >>> not windows or normalize_resource_url('file:////C:/dir/file') == 'file:///C:/dir/file'
    True
    >>> not windows or normalize_resource_url('nltk:C:/dir/file') == 'file:///C:/dir/file'
    True
    >>> not windows or normalize_resource_url('nltk:C:\\dir\\file') == 'file:///C:/dir/file'
    True
    >>> windows or normalize_resource_url('file:/dir/file/toy.cfg') == 'file:///dir/file/toy.cfg'
    True
    >>> normalize_resource_url('nltk:home/nltk')
    'nltk:home/nltk'
    >>> windows or normalize_resource_url('nltk:/home/nltk') == 'file:///home/nltk'
    True
    >>> normalize_resource_url('https://example.com/dir/file')
    'https://example.com/dir/file'
    >>> normalize_resource_url('dir/file')
    'nltk:dir/file'
    """
 ```

**Comment by ekaf** [comment]
@HyperPS, this PR still has multiple problems. 

For ex., the changes to `_BOM_TABLE['utf16']` break existing behavior and introduce bugs:

- `codecs.BOM16_BE` does not exist; the correct constant is `codecs.BOM_UTF16_BE`. Using `BOM16_BE` will raise an `AttributeError` at import time.
- The new conditional can return `{}` for the `'utf16'` entry. `_check_bom` assumes `_BOM_TABLE[enc]` is an iterable of `(bom, new_encoding)` tuples:

  ```python
  for bom, new_encoding in bom_info:
      ... 
If bom_info is a dict, this loop will misbehave or fail. Callers rely on _BOM_TABLE having the original list-of-tuples shape.

Please revert this entry to the original structure (or keep any changes within the same list-of-tuples shape).

**Comment by ekaf** [comment]
Summary of remaining issues (overview)

- _BOM_TABLE corruption
  - Ensure `_BOM_TABLE['utf16']` uses the standard `codecs.BOM_UTF16_LE` / `codecs.BOM_UTF16_BE` constants and remains a list of `(bom_bytes, new_encoding)` tuples. Do not return a dict or use non-existent names (e.g. `BOM16_BE`).

- find() lacks defense-in-depth
  - Add an explicit traversal/absolute-path rejection inside `find()` so callers that bypass `normalize_resource_url()` cannot reach filesystem paths via traversal segments.

- Unsafe regex implementation
  - Compile `_UNSAFE_NO_PROTOCOL_RE` as a raw string (r"""...""") to avoid Python string escape issues, and tighten the pattern to avoid false positives/negatives (verify backslashes, drive letters, absolute paths, and valid package paths).

- Zipfile regex greedy capture
  - Use a non-greedy or more precise pattern (e.g. `r"(.*?\.zip)/?(.*)$"` or `r"(^|.*/)([^/]+\.zip)/?(.*)$"`) so the left-most `.zip` is treated as the archive and the remainder as the entry. Add tests to assert this behavior.

- Edge-case handling for colon/protocol detection
  - Revisit the logic that skips validation when a colon is present to ensure constructs like Windows drive letters are handled safely while not bypassing checks for other malformed inputs.

- Error message / UX polish
  - Avoid embedding raw ANSI escapes or literal `{resource}` placeholders in library error strings. Use formatted messages and consider removing terminal escapes from exceptions.

Suggested unit tests to add
- BOM table integrity: shape and constants for `'utf16'`.
- normalize_resource_url rejects traversal/backslash/absolute no-protocol inputs but allows valid package paths.
- find() rejects traversal-like names when called directly.
- Zipfile regex tests: nested `.zip` paths assert first (left-most) `.zip` is captured as zipfile.
- Round-trip/integration tests covering load/retrieve for `nltk:` vs `file:` inputs.

These tests will make the remaining issues visible and prevent regressions as fixes are applied.

**Comment by ekaf** [comment]
Thanks @HyperPS! Some of the following tests still show remaining problems: 
[test_data_security.py](https://github.com/user-attachments/files/24026419/test_data_security.py)

**Comment by ekaf** [comment]
The unit tests would pass if they just expected LookupError instead of ValueError. However, it is more helpful if find() and normalize_resource_url() raise ValueError instead, so that missing-resources and invalid-inputs remain distinguishable.

Also, it is preferable to use RE.search rather than RE.match, to detect unsafe patterns anywhere in the string.

Last, you may consider replacing the current _UNSAFE_NO_PROTOCOL_RE definition with this (more comprehensive)
`_UNSAFE_NO_PROTOCOL_RE = re.compile(
     r"(?:\.\./|\.\.$|^/|\\|^[A-Za-z]:[/\\])"
)
`

**Comment by HyperPS** [comment]
Ok @ekaf i will push updates soon.

**Comment by ekaf** [comment]
The situation has improved after integrating #3484, and fixing non-Windows paths. Now only 2 failures remain, due to pytest expecting ValueError, where LookupError is raised instead.
@HyperPS, note that all your PR's now need to integrate #3484.

**Comment by ekaf** [comment]
@HyperPS, you have currently 6 PRs open, at various stages of finish. My impression is that the two oldest (this one #3467, and maybe #3468 soon) are getting ripe for merging. 
However, it is a concern that none of the problems solved by your PRs is discussed in any issue, and that we lack information about overlaps between the code additions proposed in these 6 PRs, and how these overlaps eventually could affect the recommended merge order. between the PRs.
So it would be helpful if you opened an issue providing an overview of your plans with these updates.

**Comment by HyperPS** [comment]
@ekaf, thanks for the clarification — that makes sense.

For context, all the security issues addressed by my PRs were reported privately on Huntr for coordinated disclosure after fixes. Each report includes a PoC and references the corresponding PRs.
One example is the Zip Slip vulnerability, which was validated by @purificant and has already been assigned CVE-2025-14009.
The remaining reports are still pending and were shared privately by Huntr via email to the address listed in SECURITY.md (nltk.team@gmail.com).

To align with NLTK’s preferred workflow, could you please advise on the next step for CVE assignment for these pending reports?
• should CVEs be assigned once the related PRs are merged, or
• would you prefer that I open GitHub issues summarizing each report (with links to the private Huntr disclosures), so CVE IDs can be tracked there?

I’m happy to proceed in whichever way you recommend.

**Comment by ekaf** [comment]
Thanks @HyperPS, I don't know enough about Huntr to be able to answer your CVE assignment questions. Let's hope @purificant can help you there.
What I would like to know is if there is a recommended merge order of your PRs, to avoid that code overlaps between them create conflicts.

**Comment by HyperPS** [comment]
@ekaf 
Regarding merge order: there is **no strict dependency or required sequencing** between PRs.
Each PR addresses a **distinct vulnerability in a separate area**, for example:
- `nltk.data.find()` path traversal (reported via GHSA)
- Zip Slip in the NLTK downloader (CVE assigned CVE-2025-14009)
- CorpusReader / `FileSystemPathPointer` path traversal (pending)
- StanfordSegmenter untrusted JAR loading (pending)
- Arbitrary File Read via Absolute Path Input in nltk.util.filestring() enabling Local & Remote File Disclosure  (pending)
- Arbitrary File Read via FileSystemPathPointer + PlaintextCorpusReader (bypass even if nltk.data.find() is patched ..(pending)
Any CVE assignment or grouping will depend on how @purificant prefers to handle them.

**Comment by ekaf** [comment]
@HyperPS, given the difficulty of getting #3467 and #3468 finished, I suggest trying to land these first, before eventually reviewing your remaining PRs.

File: nltk/data.py [hunk 2]
```diff
@@ -532,15 +573,22 @@ def find(resource_name, paths=None):
     :rtype: str
     """
     resource_name = normalize_resource_name(resource_name, True)
+    # Defense-in-depth: reject traversal/absolute paths even if caller bypassed normalize_resource_url()
```
[the remark this answers is not in the document]
Comment (HyperPS) [reply nltk/data.py:590]: Do we need  this extra comments in code ? @ekaf

[the remark this answers is not in the document]
Comment (ekaf) [reply nltk/data.py:590]: No, that's too much.

File: nltk/data.py [hunk 3]
```diff
@@ -46,6 +46,24 @@
 from io import BytesIO, TextIOWrapper
 from urllib.request import url2pathname, urlopen
 
+# Reject unsafe no-protocol paths: traversal segments, trailing '..', absolute paths,
+# backslashes, Windows drive letters. Use a raw-string pattern and do not anchor only
+# at the start — we'll use search() for safety checks.
+_UNSAFE_NO_PROTOCOL_RE = re.compile(r"(?:\.\./|\.\.$|^/|\\|^[A-Za-z]:[/\\])")
```
[the remark this answers is not in the document]
Comment (ekaf) [reply nltk/data.py]: ok

File: nltk/test/unit/test_data_security.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    with pytest.raises(ValueError):
+        data.normalize_resource_url("/etc/passwd")
+
+
+def test_normalize_rejects_no_protocol_windows_drive_letter_paths():
+    with pytest.raises(ValueError):
+        data.normalize_resource_url(r"C:\etc\passwd")
+    if sys.platform.startswith("win"):
+        with pytest.raises(ValueError):
+            data.normalize_resource_url("C:/etc/passwd")
```
[the remark this answers is not in the document]
Comment (ekaf) [reply nltk/test/unit/test_data_security.py]: done

File: nltk/tag/mapping.py [hunk 5]
```diff
@@ -56,7 +56,9 @@
 
 
 def _load_universal_map(fileid):
-    contents = load(join(_UNIVERSAL_DATA, fileid + ".map"), format="text")
+    base = "nltk:" + _UNIVERSAL_DATA.lstrip("/")
+    resource = f"{base.rstrip('/')}/{fileid}.map"
+    contents = load(resource, format="text")
```
[the remark this answers is not in the document]
Comment (ekaf) [reply nltk/tag/mapping.py]: ok

File: nltk/test/unit/test_data_security.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+    We don't require NLTK data to be present; we only assert the LookupError message
+    mentions the expected zipfile component and not the greedy one.
+    """
+    resource = "dir1/dir2/a.zip/b.zip/c.txt"
+    m = _match_zip_non_greedy(resource)
+    assert m is not None
+    zipfile, zipentry = m.groups()
+    assert zipfile == "dir1/dir2/a.zip"
+    assert zipentry == "b.zip/c.txt"
```
[the remark this answers is not in the document]
Comment (ekaf) [reply nltk/test/unit/test_data_security.py]: ok

File: nltk/test/unit/test_data_security.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    assert out.startswith(
+        "nltk:"
+    ), "Package-style paths should be treated as 'nltk:' URLs"
+
+
+def test_find_rejects_traversal_direct_call():
+    """Defense-in-depth: direct calls to find() should reject traversal-like names."""
+    with pytest.raises(ValueError):
+        data.find("../../etc/passwd")
+
```
[the remark this answers is not in the document]
Comment (ekaf) [reply nltk/test/unit/test_data_security.py:35]: done

**Review by ekaf — APPROVED** [review]
All concerns and suggestions seem addressed: LGTM.

**Comment by stevenbird** [comment]
Thanks for all your work @HyperPS and @ekaf!
